### PR TITLE
Ga4 frontend capability team

### PR DESCRIPTION
**Pull request by andrew-moores**
### What changed

- **Install the One Login GA4 Node Package:** Upgrade to the latest version which resolves the present bugs still needs to be done
- **Add Build Command for Analytics:** Introduce a new command (build-js:analytics) in package.json to facilitate copying the analytics file from the GA4 package into the dist/ga4/scripts folder during the build process.
- **Update Base Template with appInit Function**: Incorporate the appInit function into the base.njk file.
- **Update Config Functions**: Implement four new functions in config.js to retrieve variables (disableUa, disableGa4, UaContainerId, GA4ContainerId) from the .env file.
- **Extend Locals Middleware**: Expand the setLocalsVarsMiddleware function to include the four variables (disableUa, disableGa4, UaContainerId, GA4ContainerId), enabling access to them in the Nunjucks files.
- **Introduce GA4 OnPageLoad Component**: Create the ga4OnPageLoad Nunjucks macro component, a script tag that dynamically sets taxonomy values for the pageViewTracker on each page. This component resides in components/common/ga4-opl folder.
- **Import GA4 OnPageLoad Component**: Import the ga4OnPageLoad component into base.njk file
- **Implement GA4 OnPageLoad Component**: Add the ga4OnPageLoad Nunjucks component to all pages, incorporating relevant taxonomy value data. PageTitle, Taxonomy and contentID values configured to be set dynamically
- **Modify Cookie Banner Logic**: Adjust the cookie banner logic in the views/common/layout/banner.njk file on line 88 from "true" to "false." This change addresses a bug where the banner wasn't appearing due to our package hiding it by default upon installation.
- **Form Change Setup** - Added ?edit=true parameter to url on change links, so that FormChange Tracker is activated and fired.

**Andrew from the home team has said he will resolve any merge conflicts**

### Why did it change

To Implement the GA4 Package

### Environment variables or secrets

- [ ] Application configuration is up-to-date
- [ ] Documented in the README
- [ ] Added to deployment steps
- [ ] Added to local startup config

## Testing

- Tested OPL tracker is firing on all pages
- Tested Navigation , Form Response , Form Error and Form Change Trackers are firing and pushing the correct data where expected.

On the back of the testing, bug tickets have been raised and addressed. We are currently awaiting the release of the latest package version to update and resolve the identified issues.

Notes & Screenshots from testing [here](https://docs.google.com/document/d/11c0QAnOEYV6ZLLRYYhxvtOCbq1P-jZyJIdZRyix6Oq8/edit?pli=1)

## How to review

To evaluate the implementation, the most effective method is to test the trackers using Google Tag Manager.

[DFC-320]: https://govukverify.atlassian.net/browse/DFC-320?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/components/common/ga4-opl/template.njk [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<script {% if params.nonce%} nonce='{{ params.nonce }}'{%  endif %}>
+    window.addEventListener('load', function () {
+        window
+            .DI
```
Comment (aenciso) [inline src/components/common/ga4-opl/template.njk:4]: I think is safer to call `window.DI?.analyticsGa4` just to make sure it doesn't throw if DI is undefined

**Comment by SophiaDWokoma** [comment]
> Hi, thanks for your work on this! Not sure whether this is due to a configuration issue or something similar on my end, but I checked out the branch and noticed a "Running push to dataLayer" log message and an error in the console. The error appears for me although I haven't engaged with the cookie banner. It seems like it's breaking at `window.DI.analyticsUa.init())`, I guess what is happening is `analyticsUa` doesn't exist on the `DI` object but I haven't dug into why as I'm about to log off for the day.
> 
>  recording.mov 
> It seems like there's an attempt to run the initialisation function although cookies have not been accepted. Not sure whether that's quite right. If cookies haven't been accepted explicitly, I'm not sure that any type of tracker function should be running.

Hi Dana , I am just going to check but I think this might be addressed in the new version of the package which has some bug fixes and updates, which is due to be released. But will get back to you on this as I agree , I don't think it should be running if cookies aren't accepted

File: src/components/common/layout/base.njk [hunk 2]
```diff
@@ -127,4 +128,17 @@
         }
       {% endif %}
     </script>
+    <script type="text/javascript" src="/public/scripts/analytics.js" nonce='{{scriptNonce}}'></script>
+    <script type="text/javascript" {% if scriptNonce %} nonce="{{ scriptNonce }}"{%  endif %}>
+    window
+      .DI
```
[the remark this answers is not in the document]
Comment (SophiaDWokoma) [reply src/components/common/layout/base.njk:134]: Hello Alfonso, I agree , but will just need verify if this update aligns with our implementation standards, as we've been consistently using the same approach across all repositories.

**Comment by SophiaDWokoma** [comment]
> > Hi, thanks for your work on this! Not sure whether this is due to a configuration issue or something similar on my end, but I checked out the branch and noticed a "Running push to dataLayer" log message and an error in the console. The error appears for me although I haven't engaged with the cookie banner. It seems like it's breaking at `window.DI.analyticsUa.init())`, I guess what is happening is `analyticsUa` doesn't exist on the `DI` object but I haven't dug into why as I'm about to log off for the day.
> > recording.mov
> > It seems like there's an attempt to run the initialisation function although cookies have not been accepted. Not sure whether that's quite right. If cookies haven't been accepted explicitly, I'm not sure that any type of tracker function should be running.
> 
> Hi Dana , I am going to check but I think this might be addressed in the new version of the package which has some bug fixes and updates, which is due to be released soon. But will get back to you on this, as I agree I don't think it should be running if cookies aren't accepted. The runningPushToDataLayer is expected as it is coming from our PageView Tracker , We're still in the implementation phase at the moment and are using it as a way to verify the tag is running while we do this. We have a tech debt ticket to remove these once roll out is completed

@danacotoran Just checked with the team regarding the error in console,  yes window.DI.analyticsUa.init()) is supposed to run , even if the user hasn't engaged with the cookie banner but there shouldn't be an error. I cant see it the error however on my side.

File: src/components/common/layout/base.njk (hunk 2 above)
[the remark this answers is not in the document]
Comment (SophiaDWokoma) [reply src/components/common/layout/base.njk:134]: @aenciso I have made the change as requested , using if conditions i

**Comment by danacotoran** [comment]
@andrew-moores is this PR still needed? 👀  If not, could it be closed?

**Pull request closed without merging**